appserver/topology/graph.go

```diff
@@ -0,0 +1,124 @@
+package topology
+
+import (
+	"encoding/json"
+)
+
+// Graph contains the groupds, edges and nodes of the graph.
+type Graph struct {
+	Nodes  []Node  `json:"nodes,omitempty" protobuf:"bytes,1,opt,name=nodes"`
+	Edges  []Edge  `json:"edges,name=edges"`
+	Groups []Group `json:"groups,name=groups"`
+}
+
+// Node is the id and name of a node.
+type Node struct {
+	ID   string `json:"id,name=id"`
+	Name string `json:"name,name=name"`
+}
+
+// Edge is the source and target of a node.
+type Edge struct {
+	Source string `json:"source,name=source"`
+	Target string `json:"target,name=target"`
+}
+
+// Group is a group of nodes.
+type Group struct {
+	ID    string   `json:"id,name=id"`
+	Name  string   `json:"name,name=name"`
+	Nodes []string `json:"nodes,name=nodes"`
+}
+
+// Resource of a node.
+type Resource struct {
+	Metadata string `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	Status   string `json:"status,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	Name     string `json:"name,name=name"`
+	Kind     string `json:"kind,name=kind"`
```

Comment:
A Resource may have a `Spec` as well, so we can add that here

Reply:
fixex here: 01d64cf38e14823b83fdf3fa71fd5d243f321be5